File: Number of Good Pairs.java
/*
Time Complexity : O(n)
Space Complexity : O(n)
*/

class Solution {
	public int numIdenticalPairs(int[] nums) {

		HashMap<Integer, Integer> mp = new HashMap<Integer, Integer>() ;

		Integer n = nums.length , count = 0 ;

		for ( Integer i = 0 ; i < n ; i++ ) {
			if (mp.get(nums[i]) != null) {
				count +=  mp.get(nums[i]) ;
				mp.put(nums[i], 1 + mp.get(nums[i])) ;
			} else {
				mp.put(nums[i], 1) ;
			}
		}

		return count ;
	}
}


/*
Time Complexity : O(n^2)
Space Complexity : O(1)
*/

class Solution {
	public int numIdenticalPairs(int[] nums) {
		Integer count = 0 ;

		for ( Integer i = 0 ; i < nums.length ; i++ ) {
			for ( Integer j = i + 1 ; j < nums.length ; j++ ) {
				if ( nums[i] == nums[j] ) count++ ;
			}
		}

		return count ;
	}
}

File: Number of Good Pairs.cpp
/*
Time Complexity : O(n)
Space Complexity : O(n)
Single Scan
*/

class Solution {
public:
	int numIdenticalPairs(vector<int>& nums) {
		unordered_map<int, int> mp ;

		int n = nums.size() ;

		int count = 0 ;

		for ( int i = 0 ; i < n ; i++ ) {
			if (mp[nums[i]]) {
				count +=  mp[nums[i]] ; mp[nums[i]]++ ;
			}
			else {
				mp[nums[i]]++ ;
			}
		}

		return count ;
	}
};

/*
Time Complexity : O(n)
Space Complexity : O(n)
Two Scans
*/

class Solution {
public:
	int numIdenticalPairs(vector<int>& nums) {
		unordered_map<int, int> mp ;

		int n = nums.size() ;

		for ( int i = 0 ; i < n ; i++ ) {
			mp[nums[i]]++ ;
		}

		int count = 0 ;

		for ( auto itr = mp.begin() ; itr != mp.end() ; ++itr ) {
			int k = itr->second ;
			count += k * (k - 1) / 2 ;
		}

		return count ;
	}
};


/*
Time Complexity : O(n^2)
Space Complexity : O(1)
*/


class Solution {
public:
	int numIdenticalPairs(vector<int>& nums) {

		int count = 0 ;

		for ( int i = 0 ; i < nums.size() ; i++ ) {
			for ( int j = i + 1 ; j < nums.size() ; j++ ) {
				if (nums[i] == nums[j]) count++ ;
			}
		}
		return count ;
	}
};